src/kernel-install/kernel-install.c

```diff
@@ -0,0 +1,1085 @@
+
+        log_debug("entry-token \"%s\" acquired from %s.", token, path);
+        c->entry_token = TAKE_PTR(token);
+        return 1; /* loaded */
+}
+
+static int context_ensure_entry_token(Context *c) {
+        assert(c);
+
+        if (c->entry_token)
+                return 0;
+
+        c->entry_token = strdup(SD_ID128_TO_STRING(c->machine_id));
+        if (!c->entry_token)
+                return log_oom();
+
+        log_debug("No entry-token candidate matched, using \"%s\" from machine-id.", c->entry_token);
+        return 0;
+}
+
+static int context_load_plugins(Context *c) {
+        int r;
+
+        assert(c);
+
+        if (c->plugins)
+                return 0;
+
+        r = conf_files_list_strv(&c->plugins,
+                                 ".install",
+                                 /* root = */ NULL,
+                                 CONF_FILES_EXECUTABLE | CONF_FILES_REGULAR | CONF_FILES_FILTER_MASKED,
+                                 STRV_MAKE_CONST("/usr/lib/kernel/install.d", "/etc/kernel/install.d"));
+        if (r < 0)
+                return log_error_errno(r, "Failed to find plugins: %m");
+
+        return 0;
+}
+
+static int context_init(Context *c) {
+        int r;
+
+        assert(c);
+
+        r = context_load_environment(c);
+        if (r < 0)
+                return r;
+
+        r = context_load_install_conf(c);
+        if (r < 0)
+                return r;
+
+        r = context_load_machine_info(c);
+        if (r < 0)
+                return r;
+
+        r = context_ensure_machine_id(c);
+        if (r < 0)
+                return r;
+
+        /* Now that we determined the machine ID to use, let's determine the "token" for the boot loader
+         * entry to generate. We use that for naming the directory below $BOOT where we want to place the
+         * kernel/initrd and related resources, as well for naming the .conf boot loader spec entry.
+         * Typically this is just the machine ID, but it can be anything else, too, if we are told so. */
+
+        r = context_load_entry_token(c);
+        if (r < 0)
+                return r;
+
+        r = context_find_boot_root_and_entry_token(c);
+        if (r < 0)
+                return r;
+
+        r = context_ensure_boot_root(c);
+        if (r < 0)
+                return r;
+
+        r = context_ensure_entry_token(c);
+        if (r < 0)
+                return r;
+
+        r = context_load_plugins(c);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static int context_inspect_kernel(Context *c) {
+        assert(c);
+
+        if (!c->kernel)
+                return 0;
+
+        return inspect_kernel(c->kernel, &c->kernel_image_type, NULL, NULL, NULL);
+}
+
+static int context_ensure_layout(Context *c) {
+        _cleanup_free_ char *path = NULL, *srel = NULL;
+        int r;
+
+        assert(c);
+        assert(c->boot_root);
+        assert(c->entry_token);
+
+        if (c->layout > 0)
```

Comment:
hmm, i would write this as:

```c
if (c->layout >= 0 && c->layout != LAYOUT_AUTO)
```

it's the same but a bit more readable, since you don't have to know that LAYOUT_AUTO is zero in that case.

but doesn't matter